src/toxcore/dht_new/client.rs

```diff
@@ -0,0 +1,236 @@
+        client.send_to(tx, client.addr, packet.clone()).wait().unwrap();
+        let (received_packet, _rx) = rx.into_future().wait().unwrap();
+        assert_eq!(packet, received_packet.unwrap().1);
+    }
+    // send_ping_response()
+    #[test]
+    fn client_send_ping_response_test() {
+        let (client, sk, tx, rx) = create_client();
+        let payload = PingResponsePayload { id: random_u64() };
+        client.send_ping_response(tx, payload).wait().unwrap();
+        let (received_packet, _rx) = rx.into_future().wait().unwrap();
+        if let DhtPacket::PingResponse(packet) = received_packet.unwrap().1 {
+            let ping_resp_payload = packet.get_payload(&sk).unwrap();
+            assert_eq!(ping_resp_payload.id, payload.id);
+        } else {
+            unreachable!("can not occur");
+        }
+    }
+    // send_nodes_response()
+    #[test]
+    fn client_send_nodes_response_test() {
+        let (client, sk, tx, rx) = create_client();
+        let payload = NodesResponsePayload { nodes: vec![
+            PackedNode::new(false, SocketAddr::V4("127.0.0.1:12345".parse().unwrap()), &gen_keypair().0)
+        ], id: 38 };
+        client.send_nodes_response(tx, payload.clone()).wait().unwrap();
+        let (received_packet, _rx) = rx.into_future().wait().unwrap();
+        if let DhtPacket::NodesResponse(packet) = received_packet.unwrap().1 {
+            let nodes_resp_payload = packet.get_payload(&sk).unwrap();
+            assert_eq!(nodes_resp_payload.id, payload.id);
+            assert_eq!(nodes_resp_payload.nodes, payload.nodes);
+        } else {
+            unreachable!("can not occur");
+        }
+    }
+    // send_nat_ping_response()
+    #[test]
+    fn client_send_nat_ping_response_test() {
+        let (client, sk, tx, rx) = create_client();
+        let payload = NatPingResponse { id: random_u64() };
+        client.send_nat_ping_response(tx, &client.pk, payload).wait().unwrap();
+        let (received_packet, _rx) = rx.into_future().wait().unwrap();
+        if let DhtPacket::DhtRequest(packet) = received_packet.unwrap().1 {
+            if let DhtRequestPayload::NatPingResponse(nat_ping_resp_payload) = packet.get_payload(&sk).unwrap() {
+                assert_eq!(nat_ping_resp_payload.id, payload.id);
+            } else {
+                unreachable!("can not occur");
+            }
+        } else {
+            unreachable!("can not occur");
+        }
+    }
+    // relay_nat_ping_packet()
+    #[test]
+    fn client_relay_nat_ping_packet_test() {
+        let (client, sk, tx, rx) = create_client();
+        let nat_res = NatPingResponse { id: random_u64() };
+        let nat_payload = DhtRequestPayload::NatPingResponse(nat_res);
+        let dht_req = DhtRequest::new(&client.precomputed_key, &client.pk, &client.pk, nat_payload.clone());
+        client.relay_nat_ping_packet(tx, &client.addr, dht_req).wait().unwrap();
```

Comment:
Why relay_nat_ping_packet?

Reply:
It is used both for nat_ping_request/response.

Reply:
I still do not understand why you used `relay` word here.

Reply:
if the rpk is not same as this DHT node, it relay to proper DHT node.

Reply:
resend?

Reply:
send?

Reply:
Ok